backend/src/supervisors/supervisor.py

```diff
@@ -1,30 +1,86 @@
+    return (agent for agent in list_of_agents if agent.name == current_agent_name)
 
 
 def solve_all_tasks(tasks_json):
-    first_task = str(tasks_json["tasks"][0])
-    logging.info("solving first task: " + first_task)
-
-    agent_for_first_task = pick_agent(first_task)
-    logging.info("agent picked: " + agent_for_first_task)
-
-    stub_final_answer = "Something went wrong. The agent picked was " + agent_for_first_task + " which does not exist"
-
-    # TODO: This is a temp setup. Adjust to call an agent's pick_method method
-    if agent_for_first_task == "database_agent":
-        # will call agent here
-        stub_final_answer = "I have used database_agent to retrieve personal information about you"
-    if agent_for_first_task == "financial_advisor_agent":
-        # will call agent here
-        stub_final_answer = "I have called the fiancial_advisor_agent to aid in the first task"
-    if agent_for_first_task == "web_search_agent":
-        # will call agent here
-        stub_final_answer = "I have searched the web with web_search_agent to find data for the first task"
-    if agent_for_first_task == "unresolvable_agent":
-        # will call agent here
-        stub_final_answer = "I was unable to find an agent to solve the specific task."
-
-    stub_final_answer = stub_final_answer + ' Task was "' + first_task + '"'
-
-    return stub_final_answer
+    current_agent_name = "initial_agent"
+    history = []
+    attempts_count = 0
+    task_step = 0
+    task_total_count = len(tasks_json["tasks"])
+    logging.info("Solving all (" + str(task_total_count) + ") tasks")
+
+    # Check where we should keep iterating to call more agents
+    while still_solving_tasks(current_agent_name, attempts_count):
+        logging.info(f"attempts_count: {attempts_count}, task_step: {task_step}")
+        logging.info(f"current_agent: {current_agent_name}, history: {history}")
+
+        # Assign tasks
+        current_task = str(tasks_json["tasks"][task_step])
+
+        # If on the last task don't iterate
+        if (task_step == task_total_count-1):
+            next_task = "There is no next task - the Next Task is the final task (sorry I know this is confusing)"
+        else:
+            next_task = str(tasks_json["tasks"][task_step+1])
+
+        # Call LLM to decide next best step
+        best_next_step_json = pick_agent(current_task, next_task, list_of_agents, history)
+        current_agent_name = best_next_step_json["agent"]
+
+        # Check if we are done
+        if current_agent_name == "UnresolvableTaskAgent":
+            logging.info("UnresolvableTaskAgent called :( returning fail case")
+            return "I am sorry, but I was unable to find an answer to your question"
+        if current_agent_name == "GoalAchievedAgent":
+            logging.info("goal achieved!")
+            return best_next_step_json["thoughts"]["speak"] # TODO: Properly return answer when problem is solved
```

Comment:
Feels like this logic should be encapsulated within each specific agent. The supervisor shouldn't have knowledge of how each agent behaves.

Reply:
This logic is only in the case that _all_ tasks have been solved. No agent should be doing this return statement, only if we have completed the flow should we be concerned with it.